### DISCUSSION TITLE
ANBF? Formal spec?

### DISCUSSION THREAD

**Discussion by clemensv** [microsoft/typespec#3277] [General]
I am missing a formal specification of TypeSpec, and a template for extension specs, along with an ABNF that can be used to build alternate parsers or generators.

**Comment by bterlson** [comment]
Can you say more about what you want to accomplish? Extending TypeSpec is done via [libraries](https://typespec.io/docs/extending-typespec/basics), where you can implement e.g. decorators and custom emitters. We do have a [documented grammar](https://github.com/microsoft/typespec/blob/main/packages/spec/src/spec.emu.html), but it's important to note that TypeSpec is less of a format and more of a programming language. The type graph that emitters consume is not an AST but a data structure created from executing your TypeSpec program, so any custom implementation would have to implement these semantics as well.
